### PR TITLE
Reduce the garbage produced by GraphQL while handling the request

### PR DESCRIPTION
**Pull request by franz1981**
I've created a benchmark at https://github.com/franz1981/java-puzzles/blob/4a080cb08ada854a44877748f25a434e91b0b3a2/src/main/java/red/hat/puzzles/http/AcceptCharsetParseBenchmark.java can be run isolation.

The results using `JDK 17.0.6, OpenJDK 64-Bit Server VM, 17.0.6+10` are:
```
Benchmark                                                                                  Mode  Cnt          Score          Error   Units
AcceptCharsetParseBenchmark.parseCharsetsNoSplit                                          thrpt   10     739545.185 ±    18019.748   ops/s
AcceptCharsetParseBenchmark.parseCharsetsNoSplit:·gc.alloc.rate.norm                      thrpt   10       1344.115 ±        0.023    B/op
AcceptCharsetParseBenchmark.parseCharsetsNoSplitNoCharsetNoSemicolon                      thrpt   10  134652732.274 ± 32894368.169   ops/s
AcceptCharsetParseBenchmark.parseCharsetsNoSplitNoCharsetNoSemicolon:·gc.alloc.rate.norm  thrpt   10         ≈ 10⁻⁵                   B/op
AcceptCharsetParseBenchmark.parseCharsetsSplit                                            thrpt   10     111006.612 ±     4177.994   ops/s
AcceptCharsetParseBenchmark.parseCharsetsSplit:·gc.alloc.rate.norm                        thrpt   10      15585.345 ±        0.166    B/op
AcceptCharsetParseBenchmark.parseCharsetsSplitNoCharsetNoSemicolon                        thrpt   10  160658450.869 ± 18929886.627   ops/s
AcceptCharsetParseBenchmark.parseCharsetsSplitNoCharsetNoSemicolon:·gc.alloc.rate.norm    thrpt   10         ≈ 10⁻⁶                   B/op
```

The regression in the `*NoSplitNoCharsetNoSemicolon` case is due to a missing check for `;` presence upfront, but the overall new performance justify it (ie it's a trade-off).

The original issue has was due to the weird performance found at https://github.com/FgForrest/HttpServerEvaluationTest/issues/1

And, by moving from HTTP 2 to 1.1, the allocation flamegraph report the `String::split` on `getCharset` as an (average) hot path, which this PR aim to fix. 
It's not a game changer, and, in order to make it faster, I'm not anymore checking for `semicolon` presence nor `OWS` presence before the `charset=` searched, assuming a validator has already enforced it upfront.

### CODE REVIEW

**Comment by gsmet** [comment]
Is it really a hot path when actually executing a query? Because if it's not a game changer, I'm not sure about the added complexity?

**Comment by franz1981** [comment]
> Is it really a hot path when actually executing a query? 

Yep, it is hot, in the sense that is always hit, causing allocations for no reasons.
It's not a game changer, because there is another hot path, much more costly (and complex to solve) on Arc. I will send a fix for that one as well.
The code complexity can be improved by changing what it does and still avoiding String::split in such hot path. Look at the GC norm statistics: this version of the code is 15 times cheaper, memory wise for the heap: despite not being the dominant factor, it means much better usage of the Java heap memory (especially good for native image).

**Comment by gsmet** [comment]
Have you tried either compiling the pattern or using a StringTokenizer? Because that would be far more readable if it’s good enough.

**Comment by franz1981** [comment]
> Have you tried either compiling the pattern or using a StringTokenizer

I will happily do it and will use the benchmark to check how it compare; do you want me to turn the pr in draft in the meantime?

**Comment by gsmet** [comment]
> do you want me to turn the pr in draft in the meantime?

I did it. `String.split` is really inefficient given you end up compiling the pattern again and again so I'm pretty sure compiling the ``Pattern``s will help. As for the `StringTokenizer`, it might even be better but who knows :).

**Comment by franz1981** [comment]
> String.split is really inefficient given you end up compiling the pattern again and again so I'm pretty sure compiling the Patterns will help.

Not for single char strings use cases: there's an optimization to avoid creating any pattern, see:
https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/java/lang/String.java#L2283

That's implementing something similar to what I do, but I cannot use it for longer `regexp` that's why I have rolled my own (more compliant to the spec) - and avoid the `ArrayList` creation visible in the JDK version of the same optimized code (that's not much more readable, but hidden in the JDK layers!)

**Comment by franz1981** [comment]
@gsmet An additional commit to improve the code quality (and performance too):

https://github.com/franz1981/java-puzzles/blob/d2d60af3d0dfe7a2567807395138edcb1d1c24f5/src/main/java/red/hat/puzzles/http/AcceptCharsetParseBenchmark.java#L198 it's the method using `StringTokenizer`, feel free to send a PR with other variants you want to try.

Reporting the code here for simplicity
```java
    private static String getCharsetTokenizer(String mimeType) {
        if (mimeType != null) {
            var charsets = new StringTokenizer(mimeType, ";", false);
            while (charsets.hasMoreTokens()) {
                final String charset = charsets.nextToken().trim();
                if (charset.startsWith("charset=")) {
                    return charset.substring(8);
                }
            }
        }
        return StandardCharsets.UTF_8.name();
    }
```
The reason why is slower, reported below, is because it doesn't use any intrinsic of the JVM but perform search one char at time, while the current PR is making use of some JVM optimized method to search for the charset string.

The results are these (for the same commit) and e1cd39236f9373e79847ba7cbd5f0f09cef850f2 as well:

```
Benchmark                                                                                    Mode  Cnt          Score         Error   Units
AcceptCharsetParseBenchmark.parseCharsetsNoSplit                                            thrpt   10     952434.710 ±   21222.183   ops/s
AcceptCharsetParseBenchmark.parseCharsetsNoSplit:·gc.alloc.rate.norm                        thrpt   10         ≈ 10⁻³                  B/op
AcceptCharsetParseBenchmark.parseCharsetsNoSplitNoCharsetNoSemicolon                        thrpt   10  170452304.981 ± 2858513.808   ops/s
AcceptCharsetParseBenchmark.parseCharsetsNoSplitNoCharsetNoSemicolon:·gc.alloc.rate.norm    thrpt   10         ≈ 10⁻⁶                  B/op
AcceptCharsetParseBenchmark.parseCharsetsSplit                                              thrpt   10     108427.065 ±   11989.058   ops/s
AcceptCharsetParseBenchmark.parseCharsetsSplit:·gc.alloc.rate.norm                          thrpt   10      14073.218 ±       0.215    B/op
AcceptCharsetParseBenchmark.parseCharsetsSplitNoCharsetNoSemicolon                          thrpt   10  170762816.388 ±  949852.366   ops/s
AcceptCharsetParseBenchmark.parseCharsetsSplitNoCharsetNoSemicolon:·gc.alloc.rate.norm      thrpt   10         ≈ 10⁻⁶                  B/op
AcceptCharsetParseBenchmark.parseCharsetsTokenizer                                          thrpt   10     139832.649 ±    2679.046   ops/s
AcceptCharsetParseBenchmark.parseCharsetsTokenizer:·gc.alloc.rate.norm                      thrpt   10      11168.963 ±       0.056    B/op
AcceptCharsetParseBenchmark.parseCharsetsTokenizerNoCharsetNoSemicolon                      thrpt   10   34951788.207 ± 1482975.472   ops/s
AcceptCharsetParseBenchmark.parseCharsetsTokenizerNoCharsetNoSemicolon:·gc.alloc.rate.norm  thrpt   10         ≈ 10⁻⁵                  B/op
```
In short, the new version is 10X faster and can be (we need to decide which charsets to look-up) 0 garbage.
The `StringTokenizer` version is slightly better in term of speed then the existing code (but still slower then this PR) but they allocate similarly.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java [hunk 1]
```diff
@@ -242,15 +242,55 @@ private String getRequestAccept(RoutingContext ctx) {
...
+        final int firstSemicolonWhitespace = indexOfSemicolonOrWhitespace(mimeType, charsetValueStart);
+        final int charsetValueEnd = firstSemicolonWhitespace == -1 ? mimeType.length() : firstSemicolonWhitespace;
+        final int charsetLen = charsetValueEnd - charsetValueStart;
+        if (charsetLen == 0) {
+            return null;
+        }
+        return charsetOf(mimeType, charsetValueStart, charsetLen);
+    }
+
+    private static String charsetOf(String mimeType, int charsetValueStart, int charsetLen) {
```
Comment (franz1981) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:273]: @geoand we could add others too eg upper-case versions of the same charset(s)
I'm not aware which "common" charset names are setup on the wire for this. Maybe GraphQL folks knows?

Comment (franz1981) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:273]: So there are not typical values used other than utf-8 eg us-ascii or uppercased versions?

Comment (jmartisk) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:273]: Right, at least I'm not aware of anything

Comment (franz1981) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:277]: I've updated the code to cache just the UTF-8/utf-8 cases

eg
```
application/json;q=0.9, text/plain;q=0.5; charset=utf-8


application/json;q=0.9, text/plain;q=0.5; charset=UF-8
```

**Comment by franz1981** [comment]
@jmartisk @gsmet Let me know if the code is now readable enough and clear in what it does :+1:

**Review by jmartisk — APPROVED** [review]
